---
title: "HELIX meth-expr Cell adjusted results"
author: 
- Carlos Ruiz 

date: "`r Sys.Date()`"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
knitr::opts_knit$set(root.dir = "../")

```

# Introduction

In this report, we discuss the results of the association of methylation vs gene expression in the subcohort data of HELIX project. The results of this report were obtained by running a model adjusted for cell counts and only including autosome chromosomes.

```{r}
library(ggplot2)
library(dplyr)
library(cowplot)

load("results/MethComBatExpResidualsCellAdj/allres_simP_cpgs.Rdata")
load("results/preprocessFiles/allOverlaps.Rdata")
load("results/preprocessFiles/methyAnnotation.Rdata")
load("results/preprocessFiles/gexpAnnotation.Rdata")

## Define useful variables
sigdf <- df[df$sigPair, ]
allCpGs <- unique(df$CpG)
sigCpGs <- unique(sigdf$CpG)
sigTCs <- unique(sigdf$TC)
```


# General overview

We evaluated the association between `r length(unique(df$CpG))` CpGs and `r length(unique(df$TC))` TCs. Of those, `r length(sigCpGs)` CpGs were significantly associated with at least one gene (`r round(length(sigCpGs)/length(unique(df$CpG))*100, 2)`%) and `r length(sigTCs)` were significantly associated with at least one CpG (`r round(length(sigTCs)/length(unique(df$TC))*100, 2)`).

In total, we tested `r nrow(df)` CpG-TC pairs. In `r nrow(sigdf)` of those pairs, changes in methylation were associated with changes in gene expression (`r round(nrow(sigdf)/nrow(df)*100, 2)`%). Of all these associations, `r sum(sigdf$FC < 0)` were negative (`r round(mean(sigdf$FC < 0) * 100, 2)`%)

In general, each CpGs was associated with few TCs while each TCs was associated with several genes: 

```{r}
a <- data.frame(table(sigdf$CpG))
a$Feature <- "CpG"

b <- data.frame(table(sigdf$TC))
b$Feature <- "TC"
t <- rbind(a, b)
ggplot(t, aes(x = Freq)) + geom_histogram(binwidth = 1) + 
  ylab("Number of Features") + theme_bw() + facet_grid(~Feature) + 
  scale_x_continuous(limits = c(0, 30), name = "Feaures associated")

CpGsSum <- df %>%
  group_by(CpG) %>%
  summarise(Type = ifelse(sum(sigPair) == 0, "Non-significant",
                          ifelse(sum(sigPair) == 1, "Mono", "Multi")),
            Direction = ifelse(sum(sigPair) == 0, "Non-significant",
                               ifelse(all(FC[sigPair] > 0), "Positive", 
                                      ifelse(all(FC[sigPair] < 0), "Negative", "Both"))))


CpGs1 <- a[a$Freq == 1, "Var1"]
CpGs1m <- a[a$Freq != 1, "Var1"]

sigdf1 <- as_tibble(filter(sigdf, CpG %in% CpGs1))
sigdf1m <- as_tibble(filter(sigdf, CpG %in% CpGs1m))

sigdf1m_signSum <- sigdf1m %>% 
  group_by(CpG) %>% 
  summarize(propPos = mean(FC > 0), propNeg = mean(FC < 0))

allPos <- sigdf1m_signSum[sigdf1m_signSum$propPos == 1, "CpG", drop = TRUE]
allNeg <- sigdf1m_signSum[sigdf1m_signSum$propNeg == 1, "CpG", drop = TRUE]
posNeg <- sigdf1m_signSum$CpG[!sigdf1m_signSum$CpG %in% c(allPos, allNeg)]

```

`r sum(a$Freq == 1)` were only associated with one gene (`r round(mean(a$Freq == 1)*100, 2)`%), and we called them mono CpGs. On the other hand, we defined CpGs associated with more than 1 TC as pleiotropic CpGs. 

In mono CpGs, `r sum(sigdf1$FC > 0)` CpGs were positively associated with gene expression (`r round(mean(sigdf1$FC > 0)*100, 2)`%) and `r sum(sigdf1$FC < 0)` CpGs were negatively associated with gene expression (`r round(mean(sigdf1$FC < 0)*100, 2)`%). In multi CpGs, for `r sum(sigdf1m_signSum$propPos == 1)` CpGs the association was positive for all CpGs (`r round(mean(sigdf1m_signSum$propPos == 1)*100, 2)`%), for `r sum(sigdf1m_signSum$propNeg == 1)` CpGs the association was negative for all CpGs (`r round(mean(sigdf1m_signSum$propNeg == 1)*100, 2)`%) and `r sum(sigdf1m_signSum$propPos != 1 & sigdf1m_signSum$propNeg != 1)` CpGs were positively associated with some TCs and negatively associated with others (`r round(mean(sigdf1m_signSum$propPos != 1 & sigdf1m_signSum$propNeg != 1)*100, 2)`%). 

# CpG-TC distance

### Associated vs not-associated

We evaluated whether CpGs associated with gene expression were closer to the TCs. 

```{r}
df_comp <- inner_join(df, overDF, by = c("CpG", "TC"))
ggplot(df_comp, aes(x = Distance, fill = sigPair)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_fill_discrete(name = "", labels = c("Non-significant", "Significant"))
```

CpGs tend to be associated with expression of close TCs. We do not observe differences when CpGs are upstream or downstream. 

### Positive vs negative

```{r}
pA <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Negative")) %>%
  ggplot(aes(x = Distance, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_fill_discrete(name = "")

pB <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Negative")) %>%
  ggplot(aes(x = abs(Distance) + 1, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 10000), labels = c("10b", "1Kb", "100Kb")) + 
  scale_fill_discrete(name = "")

plot_grid(pA, pB, ncol = 2)

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Negative")) %>%
  lm(formula = abs(Distance) ~ class) %>% summary

```

CpGs associated negatively to TCs are on average 14Kb closer to the TSS than CpGs associated positively. 

### Mono vs Multi

```{r}
pA <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  ggplot(aes(x = Distance, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_fill_discrete(name = "")

pB <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  ggplot(aes(x = abs(Distance) + 1, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 10000), labels = c("10b", "1Kb", "100Kb")) + 
  scale_fill_discrete(name = "")

plot_grid(pA, pB, ncol = 2)

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  lm(formula = abs(Distance) ~ class) %>% summary
```

CpGs associated to only one TC are on average 26Kb closer to the TSS than CpGs associated to several TCs. 

### Both factors

```{r}
df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Negative"))) %>%
  ggplot(aes(x = abs(Distance) + 1, fill = AssocClass)) + geom_density(alpha=0.4) + 
  theme_bw() + facet_wrap(~CpGclass) + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 10000), labels = c("10b", "1Kb", "100Kb")) + 
  scale_fill_discrete(name = "")

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Negative"))) %>%
  lm(formula = abs(Distance) ~ CpGclass + AssocClass) %>% anova
```

When evaluting both factors at the same time, we get the same conclussions: CpGs inversely associated with gene expression that only change expression of one gene are closer to the TC than CpGs changing positively gene expression in more than one gene. 

# CpGs Enrichment

In the following sections, we tested if the CpGs associated with gene expression had different features.

## Genic/Intergenic

We first tested whether CpGs associated with gene expression were located in genic or intergenic regions.

### Associated vs not-associated

```{r}
rownames(methyAnnot) <- methyAnnot$Name
methyAnnot <- methyAnnot[allCpGs, ]
methyAnnot$assoc <- methyAnnot$Name %in% sigCpGs

methyAnnot$GeneRel <- ifelse(methyAnnot$UCSC_RefGene_Name == "", "Intergenic", "Genic")

t <- table(factor(methyAnnot$assoc, levels = c("TRUE", "FALSE")), methyAnnot$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```

CpGs associated with gene expression have an OR of `r or` of being more present in genic regions.

### Positive vs negative
```{r}
tmp <- as_tibble(methyAnnot) %>%
  mutate(CpG = Name) %>%
  left_join(CpGsSum, by = "CpG") %>%
  filter(!Direction %in% c("Non-significant", "Both")) 
t <- table(tmp$Direction, tmp$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```

CpGs negatively associated with gene expression have an OR of `r or` of being more present in genic regions than CpGs associated negatively.

### Positive vs both
```{r}
tmp <- as_tibble(methyAnnot) %>%
  mutate(CpG = Name) %>%
  left_join(CpGsSum, by = "CpG") %>%
  filter(!Direction %in% c("Non-significant", "Negative")) 
t <- table(tmp$Direction, tmp$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```


CpGs positively associated with gene expression have a similar probality of being more present in genic regions than CpGs associated positively and negatively (OR: `r or`).

### Negative vs both
```{r}
tmp <- as_tibble(methyAnnot) %>%
  mutate(CpG = Name) %>%
  left_join(CpGsSum, by = "CpG") %>%
  filter(!Direction %in% c("Non-significant", "Positive")) 
t <- table(factor(tmp$Direction, levels = c("Negative", "Both")), tmp$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```

CpGs negatively associated with gene expression are more likely to be present in genic regions than CpGs associated positively and negatively (OR: `r or`).

### Mono vs multi

```{r}
tmp <- as_tibble(methyAnnot) %>%
  mutate(CpG = Name) %>%
  left_join(CpGsSum, by = "CpG")
t <- table(tmp$Type, tmp$GeneRel)
t
chisq.test(t)

## OR
orMoMu <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
orMoAll <- round(t[1, 1]*t[3, 2]/(t[1, 2]*t[3, 1]), 2)
orMuAll <- round(t[2, 1]*t[3, 2]/(t[2, 2]*t[3, 1]), 2)
```

CpGs that change expression of one gene, are more likely to be in the genic region than CpGs affecting expression of several genes (OR: `r orMoMu`) or than CpGs not-affecting gene expression (OR: `r orMoAll`). CpGs affecting the expression of multiple TCs are less likely to be in genic regions (OR: `r orMuAll`).

