---
title: "HELIX meth-expr Cell adjusted results"
author: 
- Carlos Ruiz 

date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
      toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE)
knitr::opts_knit$set(root.dir = "../")

```

# Introduction

In this report, we discuss the results of the association of methylation vs gene expression in the subcohort data of HELIX project. The results of this report were obtained by running a model adjusted for cell counts and only including autosome chromosomes.

```{r}
library(ggplot2)
library(dplyr)
library(cowplot)
library(tidyr)

load("results/MethComBatExpResidualsCellAdj/allres_simP_cpgs.Rdata")
load("results/preprocessFiles/allOverlaps.Rdata")
load("results/preprocessFiles/methyAnnotation.Rdata")
load("results/preprocessFiles/gexpAnnotation.Rdata")

## Define useful variables
sigdf <- df[df$sigPair, ]
allCpGs <- unique(df$CpG)
sigCpGs <- unique(sigdf$CpG)
sigTCs <- unique(sigdf$TC)
```


# General overview

We evaluated the association between `r length(unique(df$CpG))` CpGs and `r length(unique(df$TC))` TCs. Of those, `r length(sigCpGs)` CpGs were significantly associated with at least one gene (`r round(length(sigCpGs)/length(unique(df$CpG))*100, 2)`%) and `r length(sigTCs)` were significantly associated with at least one CpG (`r round(length(sigTCs)/length(unique(df$TC))*100, 2)`).

In total, we tested `r nrow(df)` CpG-TC pairs. In `r nrow(sigdf)` of those pairs, changes in methylation were associated with changes in gene expression (`r round(nrow(sigdf)/nrow(df)*100, 2)`%). Of all these associations, `r sum(sigdf$FC < 0)` were negative (`r round(mean(sigdf$FC < 0) * 100, 2)`%)

In general, each CpGs was associated with few TCs while each TCs was associated with several genes: 

```{r}
a <- data.frame(table(sigdf$CpG))
a$Feature <- "CpG"

b <- data.frame(table(sigdf$TC))
b$Feature <- "TC"
t <- rbind(a, b)
ggplot(t, aes(x = Freq)) + geom_histogram(binwidth = 1) + 
  ylab("Number of Features") + theme_bw() + facet_grid(~Feature) + 
  scale_x_continuous(limits = c(0, 30), name = "Feaures associated")

CpGsSum <- df %>%
  group_by(CpG) %>%
  summarise(Type = ifelse(sum(sigPair) == 0, "Non-significant",
                          ifelse(sum(sigPair) == 1, "Mono", "Multi")),
            Direction = ifelse(sum(sigPair) == 0, "Non-significant",
                               ifelse(all(FC[sigPair] > 0), "Positive", 
                                      ifelse(all(FC[sigPair] < 0), "Negative", "Both"))))


CpGs1 <- a[a$Freq == 1, "Var1"]
CpGs1m <- a[a$Freq != 1, "Var1"]

sigdf1 <- as_tibble(filter(sigdf, CpG %in% CpGs1))
sigdf1m <- as_tibble(filter(sigdf, CpG %in% CpGs1m))

sigdf1m_signSum <- sigdf1m %>% 
  group_by(CpG) %>% 
  summarize(propPos = mean(FC > 0), propNeg = mean(FC < 0))

allPos <- sigdf1m_signSum[sigdf1m_signSum$propPos == 1, "CpG", drop = TRUE]
allNeg <- sigdf1m_signSum[sigdf1m_signSum$propNeg == 1, "CpG", drop = TRUE]
posNeg <- sigdf1m_signSum$CpG[!sigdf1m_signSum$CpG %in% c(allPos, allNeg)]

```

`r sum(a$Freq == 1)` were only associated with one gene (`r round(mean(a$Freq == 1)*100, 2)`%), and we called them mono CpGs. On the other hand, we defined CpGs associated with more than 1 TC as pleiotropic CpGs. 

In mono CpGs, `r sum(sigdf1$FC > 0)` CpGs were positively associated with gene expression (`r round(mean(sigdf1$FC > 0)*100, 2)`%) and `r sum(sigdf1$FC < 0)` CpGs were negatively associated with gene expression (`r round(mean(sigdf1$FC < 0)*100, 2)`%). In multi CpGs, for `r sum(sigdf1m_signSum$propPos == 1)` CpGs the association was positive for all CpGs (`r round(mean(sigdf1m_signSum$propPos == 1)*100, 2)`%), for `r sum(sigdf1m_signSum$propNeg == 1)` CpGs the association was negative for all CpGs (`r round(mean(sigdf1m_signSum$propNeg == 1)*100, 2)`%) and `r sum(sigdf1m_signSum$propPos != 1 & sigdf1m_signSum$propNeg != 1)` CpGs were positively associated with some TCs and negatively associated with others (`r round(mean(sigdf1m_signSum$propPos != 1 & sigdf1m_signSum$propNeg != 1)*100, 2)`%). 

# CpG-TC distance

### Associated vs not-associated

We evaluated whether CpGs associated with gene expression were closer to the TCs. 

```{r}
df_comp <- inner_join(df, overDF, by = c("CpG", "TC"))
ggplot(df_comp, aes(x = Distance, fill = sigPair)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_fill_discrete(name = "", labels = c("Non-significant", "Significant"))
```

CpGs tend to be associated with expression of close TCs. We do not observe differences when CpGs are upstream or downstream. 

### Positive vs negative

```{r}
pA <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Negative")) %>%
  ggplot(aes(x = Distance, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_fill_discrete(name = "")

pB <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Negative")) %>%
  ggplot(aes(x = abs(Distance) + 1, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 10000), labels = c("10b", "1Kb", "100Kb")) + 
  scale_fill_discrete(name = "")

plot_grid(pA, pB, nrow = 2)

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Negative")) %>%
  lm(formula = abs(Distance) ~ class) %>% summary

```

CpGs associated negatively to TCs are on average 14Kb closer to the TSS than CpGs associated positively. 

### Mono vs Multi

```{r}
pA <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  ggplot(aes(x = Distance, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_fill_discrete(name = "")

pB <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  ggplot(aes(x = abs(Distance) + 1, fill = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 10000), labels = c("10b", "1Kb", "100Kb")) + 
  scale_fill_discrete(name = "")

plot_grid(pA, pB, nrow = 2)

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  lm(formula = abs(Distance) ~ class) %>% summary
```

CpGs associated to only one TC are on average 26Kb closer to the TSS than CpGs associated to several TCs. 

### Both factors

```{r}
df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Negative"))) %>%
  ggplot(aes(x = abs(Distance) + 1, fill = AssocClass)) + geom_density(alpha=0.4) + 
  theme_bw() + facet_wrap(~CpGclass) + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 10000), labels = c("10b", "1Kb", "100Kb")) + 
  scale_fill_discrete(name = "")

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Negative"))) %>%
  lm(formula = abs(Distance) ~ CpGclass + AssocClass) %>% anova
```

When evaluting both factors at the same time, we get the same conclussions: CpGs inversely associated with gene expression that only change expression of one gene are closer to the TC than CpGs changing positively gene expression in more than one gene. 

# CpGs Enrichment

In the following sections, we tested if the CpGs associated with gene expression had different features.

## Genic/Intergenic

We first tested whether CpGs associated with gene expression were located in genic or intergenic regions.

### Associated vs not-associated

```{r}
rownames(methyAnnot) <- methyAnnot$Name
methyAnnot <- methyAnnot[allCpGs, ]
methyAnnot$assoc <- methyAnnot$Name %in% sigCpGs
methyAnnot$GeneRel <- ifelse(methyAnnot$UCSC_RefGene_Name == "", "Intergenic", "Genic")
methyAnnot <- as_tibble(methyAnnot) %>%
  mutate(CpG = Name) %>%
  left_join(CpGsSum, by = "CpG")

t <- table(factor(methyAnnot$assoc, levels = c("TRUE", "FALSE")), methyAnnot$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```

CpGs associated with gene expression have an OR of `r or` of being more present in genic regions.

### Positive vs negative
```{r}
tmp <- methyAnnot %>%
  filter(!Direction %in% c("Non-significant", "Both")) 
t <- table(tmp$Direction, tmp$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```

CpGs negatively associated with gene expression have an OR of `r or` of being more present in genic regions than CpGs associated negatively.

### Positive vs both
```{r}
tmp <- methyAnnot %>%
  filter(!Direction %in% c("Non-significant", "Negative")) 
t <- table(tmp$Direction, tmp$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```


CpGs positively associated with gene expression have a similar probality of being more present in genic regions than CpGs associated positively and negatively (OR: `r or`).

### Negative vs both
```{r}
tmp <- methyAnnot %>%
  filter(!Direction %in% c("Non-significant", "Positive")) 
t <- table(factor(tmp$Direction, levels = c("Negative", "Both")), tmp$GeneRel)
t
chisq.test(t)

## OR
or <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
```

CpGs negatively associated with gene expression are more likely to be present in genic regions than CpGs associated positively and negatively (OR: `r or`).

### Mono vs multi

```{r}
t <- table(methyAnnot$Type, methyAnnot$GeneRel)
t
chisq.test(t)

## OR
orMoMu <- round(t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1]), 2)
orMoAll <- round(t[1, 1]*t[3, 2]/(t[1, 2]*t[3, 1]), 2)
orMuAll <- round(t[2, 1]*t[3, 2]/(t[2, 2]*t[3, 1]), 2)
```

CpGs that change expression of one gene, are more likely to be in the genic region than CpGs affecting expression of several genes (OR: `r orMoMu`) or than CpGs not-affecting gene expression (OR: `r orMoAll`). CpGs affecting the expression of multiple TCs are less likely to be in genic regions (OR: `r orMuAll`).

## Position with respect to gene 

Next, we tested whether CpGs associated with gene expression were located in specific position with respect to genes.

### Associated vs not-associated

```{r}
## Create summary tibble
posSum <- methyAnnot %>%
  group_by(Type, Direction) %>%
  summarize(TSS1500in = sum(TSS1500),
            TSS1500ou = sum(!TSS1500),
            TSS200in = sum(TSS200),
            TSS200ou = sum(!TSS200),
            UTR5in = sum(UTR5),
            UTR5ou = sum(!UTR5),
            FirstExonin = sum(FirstExon),
            FirstExonou = sum(!FirstExon),
            Bodyin = sum(Body),
            Bodyou = sum(!Body),
            UTR3in = sum(UTR3),
            UTR3ou = sum(!UTR3)) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"))

## Define vars and function
gpos <- c("TSS1500", "TSS200", "UTR5", "FirstExon", "Body", "UTR3")
getOR <- function(cols, df){
  t <- data.matrix(df[, cols])
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  c(OR = or, p.val = p.val)
}

makePlot <- function(y, title){
   as_tibble(y) %>%
  mutate(Estimate = c("OR", "p.value")) %>%
  gather("Position", "Value", 1:length(gpos)) %>%
  spread(Estimate, Value) %>%
  mutate(p.val.thres = ifelse(p.value > 0.05, "P > 0.05", 
                              ifelse(p.value < 1e-3, "P < 0.001", "P < 0.05"))) %>%
  ggplot(aes(x = Position, y = OR, fill = p.val.thres)) + 
  geom_bar(stat = "identity", colour = "black") + 
  scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Position with respect gene", 
                   limits = c("TSS1500", "TSS200", "UTR5", "FirstExon", "Body", "UTR3")) +
  scale_fill_manual(name = "", values = c("white", "grey70", "black")) +
  ggtitle(title) +
  theme_bw() + theme(plot.title = element_text(hjust = 0.5))
}

tmp <- posSum %>%
  group_by(Type0) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0))


sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
  makePlot(title = "Distribution CpGs Associated vs Non-associated")

```

CpGs associated with gene expression are more frequent in TSS1500 but more rare in TSS200 and FirstExon than CpGs not associated with gene expression.


### Directions vs non-associated
```{r}
scale <- scale_y_continuous(trans = "log2", 
                            breaks = scales::trans_breaks("log2", function(x) round(2^x, 2)),
                            limits = c(0.6, 1.8))

tmp <- posSum %>%
  filter(Direction %in% c("Negative", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) 

pNeg <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
  makePlot("Distribution CpGs Negative vs Non-significant") + scale


tmp <- posSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction))

pPos <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
 makePlot("Distribution CpGs Positive vs Non-significant") + scale

tmp <- posSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) 

pBoth <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
 makePlot("Distribution CpGs Both vs Non-significant") + scale

plot_grid(pNeg, pPos, pBoth, ncol = 3)
```

CpGs associated with gene expression are more frequent in TSS1500 and less frequent in First Exon independent of the direction of the effect. Only CpGs positively associated with gene expression are less frequent in TSS200. 

### Differences between direction
```{r}
scale <- scale_y_continuous(trans = "log2", 
                            breaks = scales::trans_breaks("log2", function(x) round(2^x, 2)),
                            limits = c(0.6, 1.8))

tmp <- posSum %>%
  filter(Direction %in% c("Negative", "Positive")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) 

pNP <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
  makePlot("Distribution CpGs Negative vs Positive") + scale


tmp <- posSum %>%
  filter(Direction %in% c("Negative", "Both")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction))

pNB <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
 makePlot("Distribution CpGs Negative vs Both") + scale

tmp <- posSum %>%
  filter(Direction %in% c("Positive", "Both")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction))

pPB <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
 makePlot("Distribution CpGs Positive vs Both") + scale

plot_grid(pNP, pNB, pPB, ncol = 3)
```

When comparing between CpGs associated with gene expression, CpGs positively associated with gene expression are less frequent in TSS200. CpGs associated positively and negatively with gene expression are more frequent in TSS1500 and less frequent in gene body. 

### Mono vs Multi vs Non-significant
```{r}
scale <- scale_y_continuous(trans = "log2", 
                            breaks = scales::trans_breaks("log2", function(x) round(2^x, 2)),
                            limits = c(0.65, 1.5))

tmp <- posSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) 

p1 <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
  makePlot("Distribution CpGs Mono vs Non-significant") + scale


tmp <- posSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum))

p2 <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
 makePlot("Distribution CpGs Multi vs Non-significant") + scale

tmp <- posSum %>%
  filter(Type %in% c("Mono", "Multi")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum))

p3 <- sapply(gpos, function(x) getOR(paste0(x, c("in", "ou")), tmp)) %>%
 makePlot("Distribution CpGs Mono vs Multi") + scale

plot_grid(p1, p2, p3, ncol = 3)
```

CpGs associated with gene expression are more frequent in TSS1500, independent of how many TCs they affect. CpGs affecting one gene are less frequent in TSS200 and more frequent in gene body with respect to CpGs not affecting gene expression or changing the expression of different TCs. CpGs affecting several genes are less frequent in First Exon and Body with respect to CpGs not associated with gene expression or with CpGs associated with one TC. 

## Chromatin state

Next, we tested whether CpGs associated with gene expression were distrubted in different chromatin status.

We used a model of 15 states. First, we tested the enrichment by grouping these states by their definition in https://www.nature.com/articles/nature14248:

> The active states (associated with expressed genes) consist of active transcription start site (TSS) proximal promoter states (TssA, TssAFlnk), a transcribed state at the 5′ and 3′ end of genes showing both promoter and enhancer signatures (TxFlnk), actively transcribed states (Tx, TxWk), enhancer states (Enh, EnhG), and a state associated with zinc finger protein genes (ZNF/Rpts). The inactive states consist of constitutive heterochromatin (Het), bivalent regulatory states (TssBiv, BivFlnk, EnhBiv), repressed Polycomb states (ReprPC, ReprPCWk), and a quiescent state (Quies). 

Para esto habría que recalcular los estados previamente !!!!

### Associated vs not-associated
```{r, eval = FALSE}

chromSumGroup <- methyAnnot %>%
  group_by(Type, Direction) %>%
  summarize(TSSproxIn = sum(TssA | TssAFlnk), 
            TSSproxOut = sum(!(TssA | TssAFlnk)),
            ActTransIn = sum(Tx | TxWk), 
            ActTransOut = sum(!(Tx | TxWk)),
            
            BivRegIn = sum(TssBiv | BivFlnk | EnhBiv), 
            BivRegOut = sum(!(TssBiv | BivFlnk | EnhBiv)), 
            reprPolyIn = sum(ReprPC | ReprPCWk), 
             reprPolyOut = sum(!(ReprPC | ReprPCWk))) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"))

groups <- c("TSSprox", "ActTrans", "BivReg", "reprPoly")

## Define vars and function
getOR <- function(cols, df){
  t <- data.matrix(df[, cols])
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  c(OR = or, p.val = p.val)
}

makePlot <- function(y, title, groups){
   as_tibble(y) %>%
  mutate(Estimate = c("OR", "p.value")) %>%
  gather("Position", "Value", 1:length(groups)) %>%
  spread(Estimate, Value) %>%
  mutate(p.val.thres = ifelse(p.value > 0.05, "P > 0.05", 
                              ifelse(p.value < 1e-3, "P < 0.001", "P < 0.05"))) %>%
  ggplot(aes(x = Position, y = OR, fill = p.val.thres)) + 
  geom_bar(stat = "identity", colour = "black") + 
  scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Chromatin State Groups", 
                   limits = groups) +
  scale_fill_manual(name = "", values = c("white", "grey70", "black")) +
  ggtitle(title) +
  theme_bw() + theme(plot.title = element_text(hjust = 0.5))
}

tmp <- chromSumGroup %>%
  group_by(Type0) %>%
  select(ends_with("In"), ends_with("Out")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0))


sapply(groups, function(x) getOR(paste0(x, c("In", "Out")), tmp)) %>%
  makePlot(title = "Distribution CpGs Associated vs Non-associated", groups = groups)
```

Next, we tested the enrichment in each chromatin state. 

### Associated vs not-associated

```{r}
## Create summary tibble
sum2 <- function(x) sum(!x)

chromStates <- c("TssA", "TssAFlnk", "TxFlnk", "TxWk", "Tx", "EnhG", "Enh",
                 "ZNF.Rpts", "Het", "TssBiv", "BivFlnk", "EnhBiv", "ReprPC",
                 "ReprPCWk", "Quies")

chromSum <- methyAnnot %>%
  group_by(Type, Direction) %>%
  summarize_at(chromStates, funs(sum, sum2)) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"))

## Define vars and function
getOR <- function(cols, df){
  t <- data.matrix(df[, cols])
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  c(OR = or, p.val = p.val)
}

makePlot <- function(y, title){
   as_tibble(y) %>%
  mutate(Estimate = c("OR", "p.value")) %>%
  gather("Position", "Value", 1:length(chromStates)) %>%
  spread(Estimate, Value) %>%
  mutate(p.val.thres = ifelse(p.value > 0.05, "P > 0.05", 
                              ifelse(p.value < 1e-3, "P < 0.001", "P < 0.05")), 
         Group = factor(ifelse(Position %in% c("TssA", "TssAFlnk"), "TssProxProm",
                        ifelse(Position %in% c("Tx", "TxWk"), "ActTrans", 
                               ifelse(Position %in% c("Enh", "EnhG"), "Enhancer", 
                                      ifelse(Position %in% c("TssBiv", "BivFlnk", "EnhBiv"), "BivReg", ifelse(Position %in% c("ReprPC", "ReprPCWk"), "ReprPoly", Position))))), 
                        levels = c("TssProxProm", "TxFlnk", "ActTrans", "Enhancer", "ZNF.Rpts", "BivFlnk", "BivReg", "Het", "ReprPoly", "Quies"))) %>%
  ggplot(aes(x = Position, y = OR, fill = p.val.thres)) + 
  geom_bar(stat = "identity", colour = "black") + 
  scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Chromatin State") +
  scale_fill_manual(name = "", values = c("white", "grey70", "black")) +
  ggtitle(title) +
    facet_grid(~ Group, scales = "free") +
  theme_bw() + theme(plot.title = element_text(hjust = 0.5))
}

tmp <- chromSum %>%
  group_by(Type0) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0))


sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
  makePlot(title = "Distribution CpGs Associated vs Non-associated")
```

CpGs associated with gene expression are more frequent in proximal promoters, enhancers and a state associated with zinc finger protein genes but not with region associated with low expression, such as repressed polycomb. 


### Directions vs non-associated
```{r}
scale <- scale_y_continuous(trans = "log2", 
                            breaks = scales::trans_breaks("log2", function(x) round(2^x, 2)),
                            limits = c(0.5, 4))

tmp <- chromSum %>%
  filter(Direction %in% c("Negative", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) 

pNeg <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
  makePlot("Distribution CpGs Negative vs Non-significant") + scale


tmp <- chromSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction))

pPos <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
 makePlot("Distribution CpGs Positive vs Non-significant") + scale

tmp <- chromSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) 

pBoth <-  sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
 makePlot("Distribution CpGs Both vs Non-significant") + scale

plot_grid(pNeg, pPos, pBoth, nrow = 3)
```

CpGs associated negatively and positive and negatively are more frequent than non-associated CpGs in the same regions (Proximal promoters and Enhancers) and they are also less frequent in inactive states (repressed polycomb and quiescent). On the other hand, CpGs associated positively with gene expression were more frequent than non-associated CpGs in inactive regions and less frequent in TxFlnk.

### Differences between direction
```{r}
scale <- scale_y_continuous(trans = "log2", 
                            breaks = scales::trans_breaks("log2", function(x) round(2^x, 2)),
                            limits = c(0.3, 4))

tmp <- chromSum %>%
  filter(Direction %in% c("Negative", "Positive")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) 

pNP <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
  makePlot("Distribution CpGs Negative vs Positive") + scale


tmp <- chromSum %>%
  filter(Direction %in% c("Negative", "Both")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction))

pNB <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
 makePlot("Distribution CpGs Negative vs Both") + scale

tmp <- chromSum %>%
  filter(Direction %in% c("Positive", "Both")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction))

pPB <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
 makePlot("Distribution CpGs Positive vs Both") + scale

plot_grid(pNP, pNB, pPB, nrow = 3)
```

In concordance with previous results, we did not observe many differences between CpGs negatively associated with gene expression or CpGs associated negatively and positively with gene expression. When comparing these CpGs with those that change expression positively, the former are more frequent in active states (TssProxProm, TxFlnk, Enhancer and ZNF) and less frequent in inactive states (ReprPoly and Quies). 


### Mono vs Multi vs Non-significant
```{r}
scale <- scale_y_continuous(trans = "log2", 
                            breaks = scales::trans_breaks("log2", function(x) round(2^x, 2)),
                            limits = c(0.4, 4))

tmp <- chromSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) 

p1 <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
  makePlot("Distribution CpGs Mono vs Non-significant") + scale


tmp <- chromSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum))

p2 <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
 makePlot("Distribution CpGs Multi vs Non-significant") + scale

tmp <- chromSum %>%
  filter(Type %in% c("Mono", "Multi")) %>%
  group_by(Type) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum))

p3 <- sapply(chromStates, function(x) getOR(paste0(x, c("_sum", "_sum2")), tmp)) %>%
 makePlot("Distribution CpGs Mono vs Multi") + scale

plot_grid(p1, p2, p3, nrow = 3)
```

CpGs associated with gene expression are more frequent in the same states, independent of the number of TCs they are associated with. The main differences that we observed were that CpGs that are associated with different TCs tend to be more frequent in ZNF regions and in bivalent regulatory states. 
