---
title: "HELIX meth-expr Cell adjusted results"
author: 
- Carlos Ruiz 

date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
      toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, fig.wide = TRUE)
knitr::opts_knit$set(root.dir = "../")

```

# Introduction

In this report, we discuss the results of the association of methylation vs gene expression in the subcohort data of HELIX project. The results of this report were obtained by running a model adjusted for cell counts and only including autosome chromosomes.

```{r}
library(org.Hs.eg.db)
library(clusterProfiler)
library(ggplot2)
library(dplyr)
library(cowplot)
library(tidyr)
library(UpSetR)

load("results/MethComBatExpResidualsCellAdj/allres_simP_cpgs.Rdata")
load("results/preprocessFiles/allOverlaps.Rdata")
load("results/preprocessFiles/methyAnnotation.Rdata")
load("results/preprocessFiles/gexpAnnotation.Rdata")

## Define useful variables
sigdf <- df[df$sigPair, ]
allCpGs <- unique(df$CpG)
sigCpGs <- unique(sigdf$CpG)
sigTCs <- unique(sigdf$TC)
```


# General overview

We evaluated the association between `r length(unique(df$CpG))` CpGs and `r length(unique(df$TC))` TCs. Of those, `r length(sigCpGs)` CpGs were significantly associated with at least one gene (`r round(length(sigCpGs)/length(unique(df$CpG))*100, 2)`%) and `r length(sigTCs)` TCs were significantly associated with at least one CpG (`r round(length(sigTCs)/length(unique(df$TC))*100, 2)`%).

In total, we tested `r nrow(df)` CpG-TC pairs. In `r nrow(sigdf)` of those pairs, changes in methylation were associated with changes in gene expression (`r round(nrow(sigdf)/nrow(df)*100, 2)`%). Of all these associations, `r sum(sigdf$FC < 0)` were negative (`r round(mean(sigdf$FC < 0) * 100, 2)`%) and `r sum(sigdf$FC > 0)` were positive (`r round(mean(sigdf$FC > 0) * 100, 2)`%).

In general, each CpGs was associated with few TCs while each TCs was associated with several genes: 

```{r}
a <- data.frame(table(sigdf$CpG))
a$Feature <- "CpG"

b <- data.frame(table(sigdf$TC))
b$Feature <- "TC"
t <- rbind(a, b)
ggplot(t, aes(x = Freq)) + geom_histogram(binwidth = 1) + 
  ylab("Number of Features") + theme_bw() + facet_grid(~Feature) + 
  scale_x_continuous(limits = c(0, 30), name = "Feaures associated")

CpGsSum <- df %>%
  group_by(CpG) %>%
  summarise(Type = ifelse(sum(sigPair) == 0, "Non-significant",
                          ifelse(sum(sigPair) == 1, "Mono", "Multi")),
            Direction = ifelse(sum(sigPair) == 0, "Non-significant",
                               ifelse(all(FC[sigPair] > 0), "Positive", 
                                      ifelse(all(FC[sigPair] < 0), "Inverse", "Both"))))


CpGs1 <- a[a$Freq == 1, "Var1"]
CpGs1m <- a[a$Freq != 1, "Var1"]

sigdf1 <- as_tibble(filter(sigdf, CpG %in% CpGs1))
sigdf1m <- as_tibble(filter(sigdf, CpG %in% CpGs1m))

sigdf1m_signSum <- sigdf1m %>% 
  group_by(CpG) %>% 
  summarize(propPos = mean(FC > 0), propNeg = mean(FC < 0))

allPos <- sigdf1m_signSum[sigdf1m_signSum$propPos == 1, "CpG", drop = TRUE]
allNeg <- sigdf1m_signSum[sigdf1m_signSum$propNeg == 1, "CpG", drop = TRUE]
posNeg <- sigdf1m_signSum$CpG[!sigdf1m_signSum$CpG %in% c(allPos, allNeg)]

```

`r sum(a$Freq == 1)` were only associated with one gene (`r round(mean(a$Freq == 1)*100, 2)`%), and we called them mono CpGs. On the other hand, we defined CpGs associated with more than 1 TC as pleiotropic CpGs. 

In mono CpGs, `r sum(sigdf1$FC > 0)` CpGs were positively associated with gene expression (`r round(mean(sigdf1$FC > 0)*100, 2)`%) and `r sum(sigdf1$FC < 0)` CpGs were negatively associated with gene expression (`r round(mean(sigdf1$FC < 0)*100, 2)`%). In multi CpGs, for `r sum(sigdf1m_signSum$propPos == 1)` CpGs the association was positive for all CpGs (`r round(mean(sigdf1m_signSum$propPos == 1)*100, 2)`%), for `r sum(sigdf1m_signSum$propNeg == 1)` CpGs the association was negative for all CpGs (`r round(mean(sigdf1m_signSum$propNeg == 1)*100, 2)`%) and `r sum(sigdf1m_signSum$propPos != 1 & sigdf1m_signSum$propNeg != 1)` CpGs were positively associated with some TCs and negatively associated with others (`r round(mean(sigdf1m_signSum$propPos != 1 & sigdf1m_signSum$propNeg != 1)*100, 2)`%). 

# CpG-TC distance

### Associated vs not-associated

We evaluated whether CpGs associated with gene expression were closer to the TCs. 

```{r}
df_comp <- inner_join(df, overDF, by = c("CpG", "TC"))
ggplot(df_comp, aes(x = Distance, color = sigPair)) + geom_density() + 
  theme_bw() + 
  scale_color_discrete(name = "", labels = c("Non-significant", "Significant"))
```

CpGs tend to be associated with expression of close TCs. We do not observe differences when CpGs are upstream or downstream. 

### Positive vs Inverse

```{r}
quantiles <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Inverse")) %>%
  group_by(class) %>%
  summarize(q05 = quantile(Distance, 0.05),
            q17 = quantile(Distance, 0.17),
            q93 = quantile(Distance, 0.83),
            q95 = quantile(Distance, 0.95)) %>%
  gather("Q", "cut", 2:5) %>%
  mutate(Interval = ifelse(Q %in% c("q05", "q95"), "90%", "66%"))

pA <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Inverse")) %>%
  ggplot(aes(x = Distance, color = class)) + geom_density() + 
  theme_bw() + 
  scale_color_discrete(name = "") +
  geom_vline(data = quantiles, aes(xintercept = cut, color = class, linetype = Interval))

pB <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Inverse")) %>%
  ggplot(aes(x = abs(Distance) + 1, color = class)) + geom_density(alpha=0.4) + 
  theme_bw() + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 1e5), labels = c("10b", "1Kb", "100Kb")) + 
  scale_color_discrete(name = "")

plot_grid(pA, pB, nrow = 2)

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Inverse")) %>%
  lm(formula = abs(Distance) ~ class) %>% summary
```

CpGs associated negatively to TCs are on average 13Kb closer to the TSS than CpGs associated positively. In both cases, there are more pairs upstream than downstream. 

### Mono vs Multi

```{r}
quantiles <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  group_by(class) %>%
  summarize(q05 = quantile(Distance, 0.05),
            q17 = quantile(Distance, 0.17),
            q93 = quantile(Distance, 0.83),
            q95 = quantile(Distance, 0.95)) %>%
  gather("Q", "cut", 2:5) %>%
  mutate(Interval = ifelse(Q %in% c("q05", "q95"), "90%", "66%"))

pA <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  ggplot(aes(x = Distance, color = class)) + geom_density() + 
  theme_bw() + 
  scale_color_discrete(name = "") +
  geom_vline(data = quantiles, aes(xintercept = cut, color = class, linetype = Interval))

pB <- df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  ggplot(aes(x = abs(Distance) + 1, color = class)) + geom_density() + 
  theme_bw() + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 1e5), labels = c("10b", "1Kb", "100Kb")) + 
  scale_color_discrete(name = "")

plot_grid(pA, pB, nrow = 2)

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  lm(formula = abs(Distance) ~ class) %>% summary
```

CpGs associated to only one TC are on average 24Kb closer to the TSS than CpGs associated to several TCs. In both cases, there are more pairs upstream than downstream. 

### Both factors

```{r}
df_comp %>%
     filter(sigPair == TRUE) %>%
     mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"),
                      AssocClass = ifelse(CpG %in% posNeg, "Both",
             ifelse(FC > 0, "Positive", "Inverse"))) %>%
     ggplot(aes(x = Distance, color = AssocClass)) + geom_density(alpha=0.4) +
     theme_bw() + facet_wrap(~CpGclass) +
   scale_x_continuous(labels = c("-500 Kb", "-200 Kb", "-100 Kb", "0", "100 Kb", "250 Kb", "500 Kb"), breaks = c(-5e5, -2.5e5, -1e5, 0, 1e5, 2.5e5, 5e5)) +
  scale_color_discrete(name = "") +
theme(axis.text.x  = element_text(angle=90, vjust=0.5))

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Inverse"))) %>%
  ggplot(aes(x = abs(Distance) + 1, color = AssocClass)) + geom_density(alpha=0.4) + 
  theme_bw() + facet_wrap(~CpGclass) + 
  scale_x_log10(name = "Distance (log10)", breaks = c(10, 1000, 1e5), labels = c("10b", "1Kb", "100Kb")) + 
  scale_color_discrete(name = "")

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Inverse"))) %>%
  lm(formula = abs(Distance) ~ CpGclass + AssocClass) %>% anova
```

When evaluting both factors at the same time, we get the same conclussions: CpGs inversely associated with gene expression that only change expression of one gene are closer to the TC than CpGs changing positively gene expression in more than one gene. 


# Effect size

### Effect size vs association

We observed the general distribution of log2FC estimates of 10% increase in methylation. 

```{r}
df %>% 
  filter(sigPair == TRUE) %>%
  ggplot(aes(x = FC/10)) + geom_histogram(binwidth = 0.10) + 
  theme_bw() + 
  scale_x_continuous(breaks = c(-10, -5, -2, -1, 0, 1, 2, 5, 10))
```

Most of the effect sizes were between -1 and 1, meaning that for an increase in 10% of methylation, gene expression varies between half and twice. Nonetheless, we observed some extreme values for logFC2, where small changes in methylation would lead to great changes in gene expression. 


### Effect size vs distance

```{r}
df_comp %>% 
  filter(sigPair == TRUE) %>%
  ggplot(aes(x = Distance, y = FC/10)) + geom_point() + 
  theme_bw() + 
  scale_color_discrete(name = "")

df_comp %>% 
  filter(sigPair == TRUE) %>%
  lm(formula = (abs(FC)) ~ (abs(Distance))) %>% summary

```

CpGs that affect more strongly gene expression tend to be slightly more closer to the TC.

```{r}
df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(Dist = ifelse(Distance > 0, "Downstream", "Upstream")) %>% 
  ggplot(aes(x = FC/10, col = Dist)) + geom_density() + 
  theme_bw() + 
  scale_color_discrete(name = "")

```

The direction of the effect does not depend on the location of the CpG (downstream or upstream).

### Positive vs Inverse

```{r}
df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Inverse")) %>%
  ggplot(aes(x = log10(abs(FC/10)), color = class)) + geom_density() + 
  theme_bw() + 
  scale_color_discrete(name = "")

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(FC > 0, "Positive", "Inverse")) %>%
  lm(formula = log10(abs(FC)) ~ class) %>% summary

```

CpGs associated negatively to TCs produce higher effects on gene expression than CpGs associated positively, though the association is weak. 

### Mono vs Multi

```{r}
df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  ggplot(aes(x = log10(abs(FC/10)), color = class)) + geom_density() + 
  theme_bw() + 
  scale_color_discrete(name = "")

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(class = ifelse(CpG %in% CpGs1, "Mono", "Multi")) %>%
  lm(formula = log10(abs(FC)) ~ class) %>% summary
```

CpGs associated to one TC are more strongly associated with gene expression than CpGs associated with several TCs.

### Both factors

```{r}
df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Inverse"))) %>%
  ggplot(aes(x = log10(abs(FC)), color = AssocClass)) + geom_density() + 
  theme_bw() + facet_wrap(~CpGclass) + 
  scale_color_discrete(name = "")

df_comp %>% 
  filter(sigPair == TRUE) %>%
  mutate(CpGclass = ifelse(CpG %in% CpGs1, "Mono", "Multi"), 
         AssocClass = ifelse(CpG %in% posNeg, "Both",
                             ifelse(FC > 0, "Positive", "Inverse"))) %>%
  lm(formula = log10(abs(FC)) ~ CpGclass + AssocClass) %>% anova
```

When evaluting both factors at the same time, we get the same conclussions: CpGs inversely associated with gene expression and with different TCs have stronger effects on gene expression. 

# CpGs Enrichment

In the following sections, we tested if the CpGs associated with gene expression had different features.


## Position with respect to gene 

Next, we tested whether CpGs associated with gene expression were located in specific position with respect to genes.

```{r}
rownames(methyAnnot) <- methyAnnot$Name
methyAnnot <- methyAnnot[allCpGs, ]
methyAnnot$assoc <- methyAnnot$Name %in% sigCpGs
methyAnnot$GeneRel <- ifelse(methyAnnot$UCSC_RefGene_Name == "", "Intergenic", "Genic")
methyAnnot <- as_tibble(methyAnnot) %>%
  mutate(CpG = Name) %>%
  left_join(CpGsSum, by = "CpG")

## Create summary tibble
posSum <- methyAnnot %>%
  group_by(Type, Direction) %>%
  summarize(Intergenicin = sum(GeneRel == "Intergenic"),
            Intergenicou = sum(GeneRel != "Intergenic"),
    TSS1500in = sum(TSS1500),
            TSS1500ou = sum(!TSS1500),
            TSS200in = sum(TSS200),
            TSS200ou = sum(!TSS200),
            UTR5in = sum(UTR5),
            UTR5ou = sum(!UTR5),
            FirstExonin = sum(FirstExon),
            FirstExonou = sum(!FirstExon),
            Bodyin = sum(Body),
            Bodyou = sum(!Body),
            UTR3in = sum(UTR3),
            UTR3ou = sum(!UTR3)) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"))

## Define vars and function
gpos <- c("Intergenic", "TSS1500", "TSS200", "UTR5", "FirstExon", "Body", "UTR3")
getOR <- function(cols, df){
  t <- data.matrix(df[, cols])
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  ORl <- log(or)
  SEl <- sqrt(1/t[1, 1] + 1/t[1, 2] + 1/t[2, 1] + 1/t[2, 2])
  c(OR = or, p.val = p.val, ORm = exp(ORl - 1.96*SEl), 
    ORM = exp(ORl + 1.96*SEl))
}

g <- function(x, gpos){
  sapply(gpos, function(y) getOR(paste0(y, c("in", "ou")), df = x))
}

sig <- posSum %>%
  group_by(Type0) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0)) %>%
  g(gpos) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:7) %>%
  mutate(Type = "Significant")

neg <- posSum %>%
  filter(Direction %in% c("Inverse", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(gpos) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:7) %>%
  mutate(Type = "Inverse")

pos <- posSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction)) %>%
  g(gpos) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:7) %>%
  mutate(Type = "Positive")

both <- posSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(gpos) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:7) %>%
  mutate(Type = "Both")

rbind(sig, neg, pos, both) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Inverse", "Positive", "Both"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
  geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
  scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Position with respect gene", 
                   limits = c("TSS1500", "TSS200", "UTR5", "FirstExon", "Body", "UTR3", "Intergenic")) +
  theme_bw() 
```

CpGs associated with gene expression are more frequent in TSS1500 and less frequent in First Exon independent of the direction of the effect. Only CpGs positively associated with gene expression are less frequent in TSS200. 

```{r}
mono <- posSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(gpos) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:7) %>%
  mutate(Type = "Mono")

multi <- posSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(gpos) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:7) %>%
  mutate(Type = "Multi")


rbind(sig, mono, multi) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Mono", "Multi"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
  geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
  scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Position with respect gene", 
                   limits = c("TSS1500", "TSS200", "UTR5", "FirstExon", "Body", "UTR3", "Intergenic"))
```

CpGs associated with gene expression are more frequent in TSS1500, independent of how many TCs they affect. CpGs affecting one gene are less frequent in TSS200 and more frequent in gene body with respect to CpGs not affecting gene expression or changing the expression of different TCs. CpGs affecting several genes are less frequent in First Exon and Body with respect to CpGs not associated with gene expression or with CpGs associated with one TC. 

## Relation to CpG Island

Next, we tested whether CpGs associated with gene expression were distributed in different CpG Island positions.

```{r}
## Create summary tibble
islandStates <- c("N_Shelf", "N_Shore", "Island", "S_Shore", "S_Shelf", "OpenSea")

islSum <- methyAnnot %>%
  group_by(Type, Direction, Relation_to_Island) %>%
  summarize(n = n()) %>%
  spread(Relation_to_Island, n) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"), 
         tot = sum(Island, N_Shelf, N_Shore, OpenSea, S_Shelf, S_Shore))

## Define vars and function
getOR <- function(col, df){
  t <- data.matrix(df[, c(col, colnames(df)[ncol(df)])])
  t[, 2] <- t[, 2] - t[, 1]
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  ORl <- log(or)
  SEl <- sqrt(1/t[1, 1] + 1/t[1, 2] + 1/t[2, 1] + 1/t[2, 2])
  c(OR = or, p.val = p.val, ORm = exp(ORl - 1.96*SEl), 
    ORM = exp(ORl + 1.96*SEl))
}

g <- function(x, gpos){
  sapply(gpos, function(y) getOR(y, df = x))
}


sigI <- islSum %>%
  group_by(Type0) %>%
  select(-c(1:2)) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0)) %>%
  g(islandStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Significant")

NegI <- islSum %>%
  filter(Direction %in% c("Inverse", "Non-significant")) %>%
  group_by(Direction) %>%
  select(-starts_with("Type")) %>%
  summarize_all(list(sum))  %>%
  g(islandStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Inverse")

PosI <- islSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(-starts_with("Type")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction)) %>%
  g(islandStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Positive")

BothI <- islSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(-starts_with("Type")) %>%
  summarize_all(list(sum)) %>%
  g(islandStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Both")



rbind(sigI, NegI, PosI, BothI) %>%
  spread(par, Value) %>%
  mutate(p.val.thres = ifelse(p.val > 0.05, "P > 0.05", 
                              ifelse(p.val < 1e-3, "P < 0.001", "P < 0.05")),
         Type = factor(Type, levels = c("Significant", "Inverse", "Positive", "Both"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "CpG Island", limits  = islandStates) +
  theme_bw() 
```

CpGs associated with gene expression are more frequent shore regions and less frequent in island and open sea. We did not observe differences in the distribution in CpG islands due to the effect direction. 

```{r}
monoI <- islSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(-c(2, 9)) %>%
  summarize_all(list(sum)) %>%
  g(islandStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Mono")

multiI <- islSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(-c(2, 9)) %>%
  summarize_all(list(sum))%>%
  g(islandStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Multi")

rbind(sigI, monoI, multiI) %>%
  spread(par, Value) %>%
  mutate(p.val.thres = ifelse(p.val > 0.05, "P > 0.05", 
                              ifelse(p.val < 1e-3, "P < 0.001", "P < 0.05")),
         Type = factor(Type, levels = c("Significant", "Mono", "Multi"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "CpG Island", limits  = islandStates) +
  theme_bw() 
```

When comparing CpGs by the number of TCs they change, we do not observe any differences. 


## Chromatin state

Next, we tested whether CpGs associated with gene expression were distrubted in different chromatin status.

We used a model of 15 states. In the figures, we grouped these states by their definition in https://www.nature.com/articles/nature14248:

> The active states (associated with expressed genes) consist of active transcription start site (TSS) proximal promoter states (TssA, TssAFlnk), a transcribed state at the 5′ and 3′ end of genes showing both promoter and enhancer signatures (TxFlnk), actively transcribed states (Tx, TxWk), enhancer states (Enh, EnhG), and a state associated with zinc finger protein genes (ZNF/Rpts). The inactive states consist of constitutive heterochromatin (Het), bivalent regulatory states (TssBiv, BivFlnk, EnhBiv), repressed Polycomb states (ReprPC, ReprPCWk), and a quiescent state (Quies). 

```{r}
## Create summary tibble
sum2 <- function(x) sum(!x)

chromStates <- c("TssA", "TssAFlnk", "TxFlnk", "TxWk", "Tx", "EnhG", "Enh",
                 "ZNF.Rpts", "Het", "TssBiv", "BivFlnk", "EnhBiv", "ReprPC",
                 "ReprPCWk", "Quies")

chromSum <- methyAnnot %>%
  group_by(Type, Direction) %>%
  summarize_at(chromStates, funs(sum, sum2)) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"))

## Define vars and function
getOR <- function(cols, df){
  t <- data.matrix(df[, cols])
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  ORl <- log(or) 
  SEl <- sqrt(1/t[1, 1] + 1/t[1, 2] + 1/t[2, 1] + 1/t[2, 2])  
  c(OR = or, p.val = p.val, ORm = exp(ORl - 1.96*SEl),      ORM = exp(ORl + 1.96*SEl))
}

g <- function(x, gpos){
  sapply(gpos, function(y) getOR(paste0(y, c("_sum", "_sum2")), x))
}

sigC <- chromSum %>%
  group_by(Type0) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0)) %>%
  g(chromStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:15) %>%
  mutate(Type = "Significant")

negC <- chromSum %>%
  filter(Direction %in% c("Inverse", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  g(chromStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:15) %>%
  mutate(Type = "Inverse")

posC <- chromSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction))%>%
  g(chromStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:15) %>%
  mutate(Type = "Positive")

bothC <- chromSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  g(chromStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:15) %>%
  mutate(Type = "Both")



rbind(sigC, negC, posC, bothC) %>%
  spread(par, Value) %>%
  mutate(p.val.thres = ifelse(p.val > 0.05, "P > 0.05", 
                              ifelse(p.val < 1e-3, "P < 0.001", "P < 0.05")),
         Type = factor(Type, levels = c("Significant", "Inverse", "Positive", "Both")),
                  Group = factor(ifelse(Region %in% c("TssA", "TssAFlnk"), "TssProxProm",
                        ifelse(Region %in% c("Tx", "TxWk"), "ActTrans", 
                               ifelse(Region %in% c("Enh", "EnhG"), "Enhancer", 
                                      ifelse(Region %in% c("TssBiv", "BivFlnk", "EnhBiv"), "BivReg", ifelse(Region %in% c("ReprPC", "ReprPCWk"), "ReprPoly", Region))))), 
                        levels = c("TssProxProm", "TxFlnk", "ActTrans", "Enhancer", "ZNF.Rpts", "BivFlnk", "BivReg", "Het", "ReprPoly", "Quies"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Chromatin State") +
   facet_grid(~ Group, scales = "free", space = "free_x") +
  theme_bw() 
```

CpGs associated with gene expression are more frequent in proximal promoters, enhancers and a state associated with zinc finger protein genes but not with region associated with low expression, such as repressed polycomb. CpGs associated negatively and positive and negatively are more frequent than non-associated CpGs in the same regions (Proximal promoters and Enhancers) and they are also less frequent in inactive states (repressed polycomb and quiescent). On the other hand, CpGs associated positively with gene expression were more frequent than non-associated CpGs in inactive regions and less frequent in TxFlnk.

```{r}
monoC <- chromSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum))  %>%
  g(chromStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:15) %>%
  mutate(Type = "Mono")

multiC <- chromSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("sum"), ends_with("sum2")) %>%
  summarize_all(list(sum)) %>%
  g(chromStates) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:15) %>%
  mutate(Type = "Multi")

rbind(sigC, monoC, multiC) %>%
  spread(par, Value) %>%
  mutate(p.val.thres = ifelse(p.val > 0.05, "P > 0.05", 
                              ifelse(p.val < 1e-3, "P < 0.001", "P < 0.05")),
         Type = factor(Type, levels = c("Significant", "Mono", "Multi")),
                  Group = factor(ifelse(Region %in% c("TssA", "TssAFlnk"), "TssProxProm",
                        ifelse(Region %in% c("Tx", "TxWk"), "ActTrans", 
                               ifelse(Region %in% c("Enh", "EnhG"), "Enhancer", 
                                      ifelse(Region %in% c("TssBiv", "BivFlnk", "EnhBiv"), "BivReg", ifelse(Region %in% c("ReprPC", "ReprPCWk"), "ReprPoly", Region))))), 
                        levels = c("TssProxProm", "TxFlnk", "ActTrans", "Enhancer", "ZNF.Rpts", "BivFlnk", "BivReg", "Het", "ReprPoly", "Quies"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Chromatin State") +
   facet_grid(~ Group, scales = "free", space = "free_x") +
  theme_bw() 
```

CpGs associated with gene expression are more frequent in the same states, independent of the number of TCs they are associated with. The main differences that we observed were that CpGs that are associated with different TCs tend to be more frequent in ZNF regions and in bivalent regulatory states. 

# Illumina annotation comparison

In the following section, we will compare whether Illumina annotation can be used to predict the genes affected by a CpG. To do so, we will compare if the genes affected by the CpG are those to which the CpG is annotated in Illumina. Thus, we will restrict our comparison to those CpGs that are associated with gene expression and to those genes that are annotated with the same name in the Illumina and Affymetrix annotations.

```{r}
expAnnDF <- as_tibble(as.data.frame(expAnnot))
expAnnDF$TC <- expAnnDF$transcript_cluster_id

sigDF <- df %>%
  as_tibble() %>%
  filter(sigPair == TRUE) %>%
  select(CpG, TC, FC) %>%
  left_join(select(methyAnnot, CpG, UCSC_RefGene_Name, Type, Direction), by = "CpG") %>%
  left_join(select(expAnnDF, TC, GeneSymbol_Affy), by = "TC") %>%
  mutate(GeneAffy = strsplit(GeneSymbol_Affy, ";"))
  
methGenes <- unique(unlist(sigDF$UCSC_RefGene_Name))
expGenes <- unique(unlist(sigDF$GeneAffy))
comGenes <- intersect(methGenes, expGenes)

methSel <- sapply(sigDF$UCSC_RefGene_Name, function(x) any(x %in% comGenes))
expSel <- sapply(sigDF$GeneAffy, function(x) any(x %in% comGenes))

sigDF2 <- sigDF[methSel & expSel, ]
sigDF2
```

In total, we have `r nrow(sigDF2)` pairs where methylation is associated with gene expression and where the genes in Illumina are also present in the Affy annotation. These pairs contain `r length(unique(sigDF2$CpG))` unique CpGs and `r length(unique(sigDF2$TC))` unique TCs. These CpGs represent a `r round(length(unique(sigDF2$CpG))/length(unique(sigDF$CpG)), 2)*100` % of CpGs associated with gene expression.


```{r}
sigDF2 <- sigDF2 %>%
  mutate(match = sapply(1:nrow(sigDF2), function(i){
   grepl(paste(sigDF2[i, "UCSC_RefGene_Name"][[1]][[1]], collapse = "|"), sigDF2[i, "GeneSymbol_Affy"])})) 

sigDF2 %>%
  group_by(Direction) %>%
  summarise(prop = mean(match)) %>%
  rbind(list("Total", mean(sigDF2$match))) %>%
  ggplot(aes(x = Direction, y = prop*100)) + geom_bar(stat = "identity", fill = "white", color = "black") + 
  scale_x_discrete(name = "Direction", limits = c("Inverse", "Positive", "Both", "Total")) + 
  scale_y_continuous(limits = c(0, 100), name = "Illumina Annotation Match") + theme_bw()
```

In 80% of the detected associations, Illumina annotation matched the gene from Affymetrix array. This proportion is lower for genes that affect gene expression in both directions, probably due to their pleiotropy.

```{r}
sigDF2 %>%
  group_by(Type) %>%
  summarise(prop = mean(match)) %>%
  rbind(list("Total", mean(sigDF2$match))) %>%
  ggplot(aes(x = Type, y = prop*100)) + geom_bar(stat = "identity", fill = "white", color = "black") + 
  scale_x_discrete(name = "Type", limits = c("Mono", "Multi", "Total")) + 
  scale_y_continuous(limits = c(0, 100), name = "Illumina Annotation Match") + theme_bw()

sigDF2 %>%filter(Type == "Multi") %>%
 group_by(CpG) %>%summarize(cot = sum(match) > 0) %>% ungroup() %>% summarize(mean(cot))

```

In CpGs affecting only one gene, Illumina annotation included the affected gene in almost all cases, while in CpGs affecting several genes, Illumina annotation missed some of the associations. Nonetheless, Illumina annotation can predict at least one of the genes affected in 84% of the multi CpGs.

# Age 

Next, we tested whether CpGs associated with gene expression also changed their methylation because of age (data from: https://www.ncbi.nlm.nih.gov/pubmed/28056824)

```{r}
agedf <- read.csv2("data/DiffMethyAgeCpGs.csv", as.is = TRUE)
agedf <- agedf %>%
  as_tibble() %>%
  mutate(Dir = ifelse(as.numeric(beta8.avg) > as.numeric(beta0.avg), "Increasing", "Decreasing"), 
         CpG = ILMNID)

ageSum <- methyAnnot %>%
  select(CpG, Type, Direction) %>%
  left_join(select(agedf, Dir, CpG), by = "CpG") %>%
  mutate(Dir = ifelse(is.na(Dir), "Constant", Dir)) %>%
  group_by(Type, Direction, Dir) %>%
  summarize(n = n()) %>%
  spread(Dir, n) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"), 
         Changed = sum(Decreasing, Increasing), 
         tot = sum(Changed, Constant))

ageG <- c("Changed", "Decreasing", "Increasing")

## Define vars and function
getOR <- function(col, df){
  t <- data.matrix(df[, c(col, colnames(df)[ncol(df)])])
  t[, 2] <- t[, 2] - t[, 1]
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  ORl <- log(or)  
  SEl <- sqrt(1/t[1, 1] + 1/t[1, 2] + 1/t[2, 1] + 1/t[2, 2]) 
  c(OR = or, p.val = p.val, ORm = exp(ORl - 1.96*SEl),      ORM = exp(ORl + 1.96*SEl))
}

g <- function(x, gpos){
  sapply(gpos, function(y) getOR(y, df = x))
}

sigA <- ageSum %>%
  group_by(Type0) %>%
  select(-c(1:2)) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0)) %>%
  g(ageG) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Significant")

negA <- ageSum %>%
  filter(Direction %in% c("Inverse", "Non-significant")) %>%
  group_by(Direction) %>%
  select(-starts_with("Type")) %>%
  summarize_all(list(sum)) %>%
  g(ageG) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Inverse")


posA <- ageSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(-starts_with("Type")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction)) %>%
  g(ageG) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Positive")

bothA <- ageSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(-starts_with("Type")) %>%
  summarize_all(list(sum)) %>%
  g(ageG) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Both")



rbind(sigA, negA, posA, bothA) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Inverse", "Positive", "Both"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Change Direction") +
  theme_bw() 
```

Of the `r length(unique(agedf$ILMNID))` CpGs whose methylation change during childhood, `r length(intersect(agedf$ILMNID, sigdf$CpG))` affect gene expression (`r round(mean(agedf$ILMNID %in% sigdf$CpG)*100, 2)`%). Of the CpGs affecting gene expression, `r round(mean(sigdf$CpG %in% agedf$ILMNID)*100, 2)`% change during childhood. 

Nonetheless, CpGs associated with gene expression are more frequent in CpGs changing during childhood. Interestingly, we observe that DNA methylation of CpGs that are inversely associted with gene expression is more likely to decrease during childhood while the DNA methylation of positive CpGs tend to increase during childhood. 


```{r}
monoA <- ageSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(-c(2, 6)) %>%
  summarize_all(list(sum)) %>%
  g(ageG) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Mono")

multiA <- ageSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(-c(2, 6)) %>%
  summarize_all(list(sum))  %>%
  g(ageG) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Multi")



rbind(sigA, monoA, multiA) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Mono", "Multi"))) %>%
  ggplot(aes(x = Region, y = OR, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Change Direction") +
  theme_bw() 
```

When comparing CpGs by the number of TCs they change, we do not observe any differences. 


# Heritability 

Next, we tested whether CpGs associated with gene expression are heritable (data from: https://www.nature.com/articles/ncomms11115). Use criteria from paper: 

- High heritability: h^2^ > 0.5
- Low heritability: h^2^ < 0.2

We evaluated full heritability and SNP heritability.

(Include only CpGs present in HELIX and paper)


### Direction of association

```{r}
## Create summary tibble
h2df <- read.csv("data/HeritabilityCpGs", as.is = TRUE)
h2df <- h2df %>%
  as_tibble() %>%
  ## Remove CpGs with NA in h2
  filter(!is.na(h2_total)) %>%
  mutate(HerT = ifelse(h2_total > 0.5, "High", 
                       ifelse(h2_total < 0.2, "Low", "Medium")),
         HerSNP = ifelse(h2_SNPs > 0.5, "High", 
                       ifelse(h2_SNPs < 0.2, "Low", "Medium")),
         CpG = cgid)

herm <- methyAnnot %>%
  select(CpG, Type, Direction) %>%
  inner_join(h2df, by = "CpG") 

herm %>%
  ggplot(aes(x = Direction, y = h2_total)) + geom_boxplot() + 
  geom_hline(yintercept = c(0.2, 0.5), linetype="dashed", colour = "blue") + 
  ggtitle("Total Heritability vs CpG direction") +
  scale_x_discrete(name = "Association Direction", 
                   limits = c("Inverse", "Positive", "Both", "Non-significant")) +
  scale_y_continuous(name = "h2", limits = c(0, 1)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))


herm$Direction <- factor(herm$Direction, levels = c("Non-significant", "Inverse", "Positive", "Both"))
                         
summary(lm(h2_total ~ Direction , herm) )

herm %>%
  filter(Direction != "Non-significant") %>%
  lm(formula = h2_total ~ Direction) %>% 
  summary()
```

CpGs associated with gene expression are more heritable than CpGs not changing gene expression. Inside CpGs changing gene expression, those CpGs affecting gene expression in both directions are the most heritable and those changing negatively gene expression the least. 

```{r}
herm %>%
  ggplot(aes(x = Direction, y = h2_SNPs)) + geom_boxplot() + 
  geom_hline(yintercept = c(0.2, 0.5), linetype="dashed", colour = "blue") + 
  ggtitle("SNP Heritability vs CpG direction") +
  scale_x_discrete(name = "Association Direction", 
                   limits = c("Inverse", "Positive", "Both", "Non-significant")) +
  scale_y_continuous(name = "h2", limits = c(0, 1)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))

summary(lm(h2_SNPs ~ Direction , herm) )

herm %>%
  filter(Direction != "Non-significant") %>%
  lm(formula = h2_SNPs ~ Direction) %>% 
  summary()
```

When we consider SNPs heritability, we still see the same effect, although the heritability is decreased. With SNP heritability, CpGs affecting gene expression in both directions are more heritable than the other CpGs affecting gene expression. 

### CpGs affected
```{r}
herm %>%
  ggplot(aes(x = Type, y = h2_total)) + geom_boxplot() + 
  geom_hline(yintercept = c(0.2, 0.5), linetype="dashed", colour = "blue") + 
  ggtitle("Total Heritability vs CpG direction") +
  scale_x_discrete(name = "Association Direction", 
                   limits = c("Mono", "Multi", "Non-significant")) +
  scale_y_continuous(name = "h2", limits = c(0, 1)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))


herm$Type <- factor(herm$Type, levels = c("Non-significant", "Mono", "Multi"))
                         
summary(lm(h2_total ~ Type , herm) )

herm %>%
  filter(Type != "Non-significant") %>%
  lm(formula = h2_total ~ Type) %>% 
  summary()
```

CpGs associated with multiple genes are more heritable than CpGs changing only one gene. 

```{r}
herm %>%
  ggplot(aes(x = Type, y = h2_SNPs)) + geom_boxplot() + 
  geom_hline(yintercept = c(0.2, 0.5), linetype="dashed", colour = "blue") + 
  ggtitle("Total Heritability vs CpG direction") +
  scale_x_discrete(name = "Association Direction", 
                   limits = c("Mono", "Multi", "Non-significant")) +
  scale_y_continuous(name = "h2", limits = c(0, 1)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))


summary(lm(h2_SNPs ~ Type , herm) )

herm %>%
  filter(Type != "Non-significant") %>%
  lm(formula = h2_SNPs ~ Type) %>% 
  summary()
```

When we use SNPs heritability, we have a higher effect.

# EWAS catalogue 

Next, we tested whether CpGs associated with gene expression are also associated with diseases or exposures in EWAS catalogue (data from: http://www.ewascatalog.org/)

```{r}
ewasdf <- read.delim(gzfile("data/EWAS_Catalog_23-11-2018.txt.gz"))

ewasdf <- mutate(ewasdf, Category = ifelse(Outcome %in% c("DNA methylation", "Changes in DNA methylation"), "Exposure", "Outcome"))

ewasSum <- methyAnnot %>%
  select(CpG, Type, Direction) %>%
  mutate(Changed = CpG %in% ewasdf$CpG,
         Exposure = CpG %in% filter(ewasdf, Category == "Exposure")$CpG, 
         Outcome = CpG %in% filter(ewasdf, Category == "Outcome")$CpG) %>%
  group_by(Type, Direction) %>%
  summarize(Changedin = sum(Changed),
            Changedou = sum(!Changed),
            Exposurein = sum(Exposure),
            Exposureou = sum(!Exposure),
            Outcomein = sum(Outcome),
            Outcomeou = sum(!Outcome)) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"))
eff <- c("Changed", "Exposure", "Outcome")

## Define vars and function
getOR <- function(cols, df){
  t <- data.matrix(df[, cols])
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  ORl <- log(or)  
  SEl <- sqrt(1/t[1, 1] + 1/t[1, 2] + 1/t[2, 1] + 1/t[2, 2]) 
  c(OR = or, p.val = p.val, ORm = exp(ORl - 1.96*SEl),      ORM = exp(ORl + 1.96*SEl))
}

g <- function(x, gpos){
  sapply(gpos, function(y) getOR(paste0(y, c("in", "ou")), df = x))
}

sigE <- ewasSum %>%
  group_by(Type0) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Significant")


negE <- ewasSum %>%
  filter(Direction %in% c("Inverse", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Inverse")

posE <- ewasSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Positive")

bothE <- ewasSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Both")

rbind(sigE, negE, posE, bothE) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Inverse", "Positive", "Both"))) %>%
  ggplot(aes(x = Region, y = OR, col = p.val.thres, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Effect in EWAS catalogue", 
                   limits = eff) +
  theme_bw() 

```

Of the `r length(unique(ewasdf$CpG))` CpGs in EWAS catalogue, `r length(intersect(ewasdf$CpG, sigdf$CpG))` affect gene expression (`r round(mean(unique(ewasdf$CpG) %in% sigdf$CpG)*100, 2)`%).  Of the CpGs affecting gene expression, `r round(mean(unique(sigdf$CpG) %in% ewasdf$CpG)*100, 2)`% are in EWAS catalogue.

CpGs associated with gene expression are more frequently found in ewas Catalogue. Of those, we observe that CpGs associated with gene expression are more typically changed by exposures than by outcomes. Nonetheless, it should be taken into account that ewas catalogue contains much more associations with exposures than with outcomes. We observe that CpGs changing gene expression in both directions are less frequently associated with an outcome. Nonetheless, the low number of CpGs associated with outcomes hinders the significances of the results. 

```{r}
monoE <- ewasSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Mono")

multiE <- ewasSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:3) %>%
  mutate(Type = "Multi")


rbind(sigE, monoE, multiE) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Mono", "Multi"))) %>%
  ggplot(aes(x = Region, y = OR, col = p.val.thres, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Effect in EWAS catalogue", 
                   limits = eff) +
  theme_bw() 

```

When comparing CpGs by the number of TCs they change, we do not observe any differences. 

# EWAS atlas

Next, we tested whether CpGs associated with gene expression are also associated with diseases or exposures in EWAS Atlas (data from: http://bigd.big.ac.cn/ewas/index/).

```{r}
atAssoc <- read.delim("data/EWAS_Atlas_associations.tsv")
atStudy <- read.delim("data/EWAS_Atlas_studies.tsv", header = FALSE)
atCohort <- read.delim("data/EWAS_Atlas_cohorts.tsv", header = FALSE)

traits <- read.csv("data/traitsTable.csv")

colnames(atStudy)[1] <- "Study.id"
colnames(atCohort)[c(2, 4, 11, 15)] <- c("Study.id", "Array", "Source", "Ancestry")
atlasDF <- atStudy %>%
  select(Study.id) %>%
  full_join(select(atCohort, Study.id, Array, Source, Ancestry), by = "Study.id") %>% 
  left_join(atAssoc, ., by ="Study.id") %>%
  as_tibble() %>%
  distinct()

## Select Europeans and blood and add trait type
atlasDFsel <- atlasDF %>%
  filter(Ancestry %in% c("Not reported", "European") &
           Source %in% c("whole blood", "peripheral blood")) %>%
  left_join(select(traits, Trait, Type), by = "Trait")
```

Of the `r length(unique(atlasDFsel$Probe.id))` CpGs in EWAS atlas, `r length(intersect(atlasDFsel$Probe.id, sigdf$CpG))` affect gene expression (`r round(mean(unique(atlasDFsel$Probe.id) %in% sigdf$CpG)*100, 2)`%).  Of the CpGs affecting gene expression, `r round(mean(unique(sigdf$CpG) %in% atlasDFsel$Probe.id)*100, 2)`% are in EWAS atlas.


```{r}
atlasSum <- methyAnnot %>%
  select(CpG, Type, Direction) %>%
  mutate(Changed = CpG %in% atlasDFsel$Probe.id,
         Behavior = CpG %in% filter(atlasDFsel, Type == "behavior")$Probe.id, 
         Cancer = CpG %in% filter(atlasDFsel, Type == "cancer")$Probe.id,
         Environment = CpG %in% filter(atlasDFsel, Type == "environmental factor")$Probe.id, 
         Disease = CpG %in% filter(atlasDFsel, Type == "non-cancer disease")$Probe.id, 
         Phenotype = CpG %in% filter(atlasDFsel, Type == "phenotype")$Probe.id) %>%
  group_by(Type, Direction) %>%
  summarize(Changedin = sum(Changed),
            Changedou = sum(!Changed),
            Cancerin = sum(Cancer),
            Cancerou = sum(!Cancer),
            Environmentin = sum(Environment),
            Environmentou = sum(!Environment),
            Behaviorin = sum(Behavior),
            Behaviorou = sum(!Behavior),
            Diseasein = sum(Disease),
            Diseaseou = sum(!Disease),
            Phenotypein = sum(Phenotype),
            Phenotypeou = sum(!Phenotype)) %>%
  mutate(Type0 = ifelse(Type == "Non-significant", "Non-significant", "Significant"))
eff <- c("Changed", "Behavior", "Disease", "Environment", "Cancer", "Phenotype")



## Define vars and function
getOR <- function(cols, df){
  t <- data.matrix(df[, cols])
  or <- t[1, 1]*t[2, 2]/(t[1, 2]*t[2, 1])
  p.val <- chisq.test(t)$p.value
  ORl <- log(or)
  SEl <- sqrt(1/t[1, 1] + 1/t[1, 2] + 1/t[2, 1] + 1/t[2, 2])
  c(OR = or, p.val = p.val, ORm = exp(ORl - 1.96*SEl),      ORM = exp(ORl + 1.96*SEl))
}

g <- function(x, gpos){
  sapply(gpos, function(y) getOR(paste0(y, c("in", "ou")), df = x))
}

sigAt <- atlasSum %>%
  group_by(Type0) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Type0)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Significant")


negAt <- atlasSum %>%
  filter(Direction %in% c("Inverse", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Inverse")

posAt <- atlasSum %>%
  filter(Direction %in% c("Positive", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  arrange(desc(Direction)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Positive")

bothAt <- atlasSum %>%
  filter(Direction %in% c("Both", "Non-significant")) %>%
  group_by(Direction) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Both")

rbind(sigAt, negAt, posAt, bothAt) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Inverse", "Positive", "Both"))) %>%
  ggplot(aes(x = Region, y = OR, col = p.val.thres, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Effect in EWAS Atlas", 
                   limits = eff) +
  theme_bw() 
```

CpGs associated with gene expression are more frequently found in ewas Atlas. Of those, we observe that CpGs associated with gene expression are more typically changed by behavioural traits and non-cancer diseases. We observe lower effects for CpGs affected by environment, specially for CpGs affecting positively or in both directions. 

```{r}
monoAt <- atlasSum %>%
  filter(Type %in% c("Mono", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Mono")

multiAt <- atlasSum %>%
  filter(Type %in% c("Multi", "Non-significant")) %>%
  group_by(Type) %>%
  select(ends_with("in"), ends_with("ou")) %>%
  summarize_all(list(sum)) %>%
  g(eff) %>%
  as_tibble() %>%
  mutate(par = c("OR", "p.val", "ORlow", "ORhigh")) %>%
  gather("Region", "Value", 1:6) %>%
  mutate(Type = "Multi")


rbind(sigAt, monoAt, multiAt) %>%
  spread(par, Value) %>%
  mutate(Type = factor(Type, levels = c("Significant", "Mono", "Multi"))) %>%
  ggplot(aes(x = Region, y = OR, col = p.val.thres, fill = Type)) + 
  geom_bar(stat = "identity", position=position_dodge()) + 
    geom_errorbar(position=position_dodge(.9), width=.25, aes(ymin = ORlow, ymax = ORhigh)) +
scale_y_continuous(trans = "log2", 
                     breaks = scales::trans_breaks("log2", function(x) round(2^x, 2))) +
  geom_hline(yintercept = 1) +
  scale_x_discrete(name = "Effect in EWAS catalogue", 
                   limits = eff)+
  theme_bw() 

```

We do not observe any differences when comparing CpGs by the number of TCs they affect. 

# Gene enrichment

In this section, we run enrichment analyses with the genes affected by the CpGs in the previous sections.

## Overall

First, we tested whether the genes regulated by DNA methylation are enriched in some pathways.

```{r}
## Make universe
allGenes <- expAnnDF %>%
  inner_join(df, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
allGenesENT <- bitr(allGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

selGenes <- expAnnDF %>%
  inner_join(filter(df, sigPair == TRUE), by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
selGenesENT <- bitr(selGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

showEnrichments <- function(genelist){
  goAll <- enrichGO(genelist, universe = allGenesENT$ENTREZID, 
                     ont = "BP",
                    OrgDb = org.Hs.eg.db)

  print(data.frame(goAll)[, 2:6])
  keggAll <- enrichKEGG(genelist,
                     universe = allGenesENT$ENTREZID, 
                     organism = "hsa")
  print(data.frame(keggAll)[, 2:6])
  return(list(go = goAll, kegg = keggAll))
}
enAll <-showEnrichments(selGenesENT$ENTREZID)
```

Genes affected by methylation are enriched in GO terms and KEGG pathways related with the immune system. 

We next tested in what pathways are enriched the genes affected by the different types of CpGs.

## Direction

```{r}
negGenes <- methyAnnot %>%
  filter(Direction == "Inverse") %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
negGenesENT <- bitr(negGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")


posGenes <- methyAnnot %>%
  filter(Direction == "Positive") %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
posGenesENT <- bitr(posGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

bothGenes <- methyAnnot %>%
  filter(Direction == "Both") %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
bothGenesENT <- bitr(bothGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

upset(fromList(list(Inverse = negGenes, Positive = posGenes,
                    Both = bothGenes)), order.by = "freq")

## Inverse
enNeg <- showEnrichments(negGenesENT$ENTREZID)

## Positive
enPos <- showEnrichments(posGenesENT$ENTREZID)

## Both
enBoth <- showEnrichments(bothGenesENT$ENTREZID)
```

CpGs regulating negatively gene expression affect to a higher number of genes. Thus, a high number of the genes affected by negative CpGs are exclusive of this group. Positive and both CpGs affect genes that are shared with other groups. 

In general, all type of CpGs regulate genes associated with immune function. Genes regulated by CpG with negative effect are involved in immune cells activation. Genes regulated by CpG with positive effect are involved in antigen presentation. 

## Type

```{r}
monoGenes <- methyAnnot %>%
  filter(Type == "Mono") %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
monoGenesENT <- bitr(monoGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")


multiGenes <- methyAnnot %>%
  filter(Type == "Multi") %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
multiGenesENT <- bitr(multiGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

upset(fromList(list(Mono = monoGenes, Multi = multiGenes)),
      order.by = "freq")

## Mono
enMono <- showEnrichments(monoGenesENT$ENTREZID)

## Multi
enMulti <- showEnrichments(multiGenesENT$ENTREZID)
```

CpGs regulating the expression of one gene affect to a higher number of genes. We observe little overlap between the genes regulated by these different types of CpGs. 

CpGs regulating one genes regulate genes involved in innate immunity while CpGs regulating several genes regulate genes involved in adapted immunity. 


## Age effect

```{r}
changeGenes <- methyAnnot %>%
  filter(CpG %in% agedf$ILMNID) %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
changeGenesENT <- bitr(changeGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

incGenes <- methyAnnot %>%
  filter(CpG %in% filter(agedf, Dir == "Increasing")$ILMNID)  %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
incGenesENT <- bitr(incGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

decGenes <- methyAnnot %>%
  filter(CpG %in% filter(agedf, Dir == "Decreasing")$ILMNID)  %>%
  select(CpG) %>%
  inner_join(filter(df, sigPair == TRUE), by = "CpG") %>%
  inner_join(expAnnDF, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
decGenesENT <- bitr(decGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")


upset(fromList(list(Decreasing = decGenes, Increasing = incGenes)), 
      order.by = "freq")

## Changed
enChanged <- showEnrichments(changeGenesENT$ENTREZID)

## Increasing
enInc <- showEnrichments(incGenesENT$ENTREZID)

## Decreasing
enDec <- showEnrichments(decGenesENT$ENTREZID)

```

CpGs that decrease their methylation during childhood regulates more genes than CpGs that increase their methylation. There is very few overlap between these set of genes. 

CpGs that changed during childhood regulate genes involved in adapted immunity. This effect is only observed for CpGs that decrease their methylation during childhood. 