---
title: "Comparison methExprs models"
author: "Carlos Ruiz"
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
      toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, fig.wide = TRUE)
knitr::opts_knit$set(root.dir = "../")
```


```{r load}
library(org.Hs.eg.db)
library(clusterProfiler)
library(dplyr)
library(ggplot2)
library(UpSetR)

load("results/preprocessFiles/gexpAnnotation.Rdata")

load("results/MethComBatExpResidualsNoCellAdj/allres_simP_cpgs.Rdata")
modU <- df
featsU <- featStatsDF

load("results/MethComBatExpResidualsCellAdj/allres_simP_cpgs.Rdata")
modC <- df
featsC <- featStatsDF

```

We fit two models. The difference is whether we included cell composition in the models:
- Model U: model adjusted for sex, cohort and age
- Model C: model adjusted for sex, cohort, age and cell composition.

# Compare CpGs

```{r, dev='Cairo_png'}
sigFU <- subset(featsU, p.val.adj < 0.05)$feat
sigFC <- subset(featsC, p.val.adj < 0.05)$feat

upset(fromList(list(Adjusted = sigFU, "Adjusted+Cells" = sigFC)), 
      order.by = "freq")
```

We found `r length(sigFU)` CpGs affecting gene expression with model U and `r length(sigFC)` with model C. Of those, `r length(sigCom)`, were found with both models. Thus, we had higher power to detect associations with model U. 

# Compare pairs

## Merge datasets

We will create a merged dataset with the estimates for both models. We will restrict the comparison to those CpG-TC pairs found significant in at least one of the models. 

```{r}
mergeTB <- modU %>%
  left_join(modC, by = c("CpG", "TC")) %>%
  as_tibble() %>%
  filter(sigPair.x == TRUE | sigPair.y == TRUE) %>%
  mutate(sigType = ifelse(sigPair.x == TRUE, ifelse(sigPair.y == TRUE, "Both", "Crude"), "Cell"))
```

In total, `r nrow(mergeTB)` CpG-TC pairs were found significant with at least any of the models. Of those, `r sum(mergeTB$sigType == "Both")` pairs were found with both models, `r sum(mergeTB$sigType == "Cell")` pairs only with the model adjusted for cell counts and `r sum(mergeTB$sigType == "Crude")` pairs only with the model not-adjusted for cell counts. 


## Compare p-values
```{r}
ggplot(mergeTB, aes(x = -log10(p.value.y), y = -log10(p.value.x), col = sigType)) +
  geom_point() +
  scale_x_continuous(name = "Cell adjusted") + 
  scale_y_continuous("Not cell adjusted") + 
  ggtitle("-log10 p-values comparative") +
  theme_bw() + theme(plot.title = element_text(hjust = 0.5)) +
  scale_color_discrete(name = "")
```

We observe some CpG-TC pairs that are only significant when we adjust for cell counts and many other that are only significant without adjusting. 

```{r}
par(mfrow = c(1, 2))
hist(modU$p.value, main = "Model U", xlab = "Raw p-values")
hist(modC$p.value, main = "Model C", xlab = "Raw p-values")
```

We obtain in general a good distribution of p-values. 


## Compare Estimates

Use this formula to estimate p-value of difference

https://stats.stackexchange.com/questions/55501/test-a-significant-difference-between-two-slope-values


```{r}
mergeTB <- mergeTB %>%
  mutate(Diff.pval = pnorm(abs(FC.x - FC.y)/sqrt(SD.x**2 + SD.y**2), lower.tail = FALSE), 
         isDiff = ifelse(Diff.pval < 0.05, "Different", "Equal"))

table(Different = mergeTB$isDiff, type = mergeTB$sigType)

ggplot(mergeTB, aes(x = FC.y, y = FC.x, col = sigType)) +
  geom_point() +
  scale_x_continuous(name = "Cell adjusted") + 
  scale_y_continuous("Not cell adjusted") + 
  ggtitle("Estimates comparative") +
  theme_bw() + theme(plot.title = element_text(hjust = 0.5)) +
  scale_color_discrete(name = "")
```

Most of the significant pairs found by the two models have similar estimates. Of those only significant in model C, a great proportion had similar estimates with model U, although they were not found as significant with this model. On the other hand, most pairs found with model U had different estimates in model C.

# Compare TCs

Finally, we compared whether the genes only found affected in one of the models were enriched in some pathways.

```{r}
expAnnDF <- as_tibble(as.data.frame(expAnnot))
expAnnDF$TC <- expAnnDF$transcript_cluster_id

## Make universe
allGenes <- expAnnDF %>%
  inner_join(modU, by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()
allGenesENT <- bitr(allGenes, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

uGenes <- expAnnDF %>%
  inner_join(filter(modU, sigPair == TRUE), by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()

cGenes <- expAnnDF %>%
  inner_join(filter(modC, sigPair == TRUE), by = "TC") %>%
  `$`("GeneSymbol_Affy") %>%
  strsplit(., split = ";") %>%
  unlist() %>%
  unique()

uGenessp <- uGenes[!uGenes %in% cGenes]
cGenessp <- cGenes[!cGenes %in% uGenes]

uGenesENT <- bitr(uGenessp, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")

cGenesENT <- bitr(cGenessp, fromType="SYMBOL", toType = "ENTREZID", OrgDb="org.Hs.eg.db")


showEnrichments <- function(genelist){
  goAll <- enrichGO(genelist, universe = allGenesENT$ENTREZID, 
                     ont = "BP",
                    OrgDb = org.Hs.eg.db)

  print(data.frame(goAll)[, 2:6])
  keggAll <- enrichKEGG(genelist,
                     universe = allGenesENT$ENTREZID, 
                     organism = "hsa")
  print(data.frame(keggAll)[, 2:6])
  return(list(go = goAll, kegg = keggAll))
}
enU <- showEnrichments(uGenesENT$ENTREZID)
enC <- showEnrichments(cGenesENT$ENTREZID)


tcU <- unique(filter(modU, sigPair == TRUE)$TC)
tcC <- unique(filter(modC, sigPair == TRUE)$TC)
```

In model C, we found `r length(tcC)` TCs whose expression was affected by DNA methylation. This number raised `r length(tcU)` in model U. Of these TCs, `r length(intersect(tcC, tcU))` were found with both models. 

Genes specific of model U are enriched in processes related with leukocyte differentiation, while genes specific of model C are not enriched in any process.
